mempool/src/pool/reorg.rs

```diff
@@ -0,0 +1,162 @@
+
+//! Support for updating the mempool upon a reorg
+
+use std::{collections::BTreeSet, mem};
+
+use chainstate::chainstate_interface::ChainstateInterface;
+use common::{
+    chain::{Block, GenBlock, SignedTransaction},
+    primitives::{Id, Idable},
+};
+use logging::log;
+use utils::{shallow_clone::ShallowClone, tap_error_log::LogError};
+use utxo::UtxosStorageRead;
+
+use crate::{
+    get_memory_usage::GetMemoryUsage,
+    pool::{tx_verifier, Mempool, MempoolStore},
+};
+
+/// An error that can happen in mempool on chain reorg
+#[derive(Debug, thiserror::Error, PartialEq, Eq)]
+pub enum ReorgError {
+    #[error(transparent)]
+    Chainstate(#[from] chainstate::ChainstateError),
+    #[error("Could not find the previous tip")]
+    OldTip,
+    #[error("Could not find the previous tip index")]
+    OldTipIndex,
+    #[error("Could not find the new tip index")]
+    NewTipIndex,
+    #[error("Block {0:?} not found while traversing history")]
+    BlockNotFound(Id<Block>),
+    #[error("Chainstate call: {0}")]
+    ChainstateCall(#[from] subsystem::subsystem::CallError),
+}
+
+/// Iterate over blocks following the parent link. Apply given function to each block.
+fn for_each_block(
+    chainstate: &dyn ChainstateInterface,
+    mut curr_id: Id<Block>,
+    stop_id: Id<GenBlock>,
+    mut func: impl FnMut(Block),
+) -> Result<(), ReorgError> {
+    let chain_config = chainstate.get_chain_config();
+    while curr_id != stop_id {
+        let block = chainstate
+            .get_block(curr_id)?
+            .ok_or_else(|| ReorgError::BlockNotFound(curr_id))?;
+        curr_id = match block.prev_block_id().classify(chain_config) {
+            common::chain::GenBlockId::Genesis(_) => break,
+            common::chain::GenBlockId::Block(block_id) => block_id,
+        };
+        func(block);
+    }
+    Ok(())
+}
+
+fn find_disconnected_txs(
+    chainstate: &dyn ChainstateInterface,
+    old_tip_id: Id<GenBlock>,
+    new_tip_id: Id<Block>,
+) -> Result<Vec<SignedTransaction>, ReorgError> {
+    let chain_config = chainstate.get_chain_config();
+    let old_tip_id = match old_tip_id.classify(chain_config) {
+        common::chain::GenBlockId::Genesis(_) => return Ok(Vec::new()),
+        common::chain::GenBlockId::Block(id) => id,
+    };
+
+    let common_id = {
+        let old_index = chainstate.get_block_index(&old_tip_id)?.ok_or(ReorgError::OldTipIndex)?;
+        let new_index = chainstate.get_block_index(&new_tip_id)?.ok_or(ReorgError::NewTipIndex)?;
+        let common_index = chainstate.last_common_ancestor(&old_index.into(), &new_index.into())?;
+        common_index.block_id()
+    };
+
+    // Short circuit the expensive processing below if no blocks are being disconnected
+    if old_tip_id == common_id {
+        return Ok(Vec::new());
+    }
+
+    // Collect IDs of trancsactions included in the newly connected chain
+    let mut connected_txs = BTreeSet::new();
+    for_each_block(chainstate, new_tip_id, common_id, |block| {
```

Comment:
This is wrong for two reasons:

1. The order of transactions matters within a single block, because transactions can have outputs in the same block. Comingling transactions in a BTreeSet will certainly not yield the correct order.
2. The blocks that are being connected should be done in an ascending order, being later blocks should be able to spend earlier blocks

Reply:
There should be tests that proves that such cases are covered.

Reply:
I'll delete this comment and rethink it as I may be wrong with what's going on here.